source/server/drain_manager_impl.cc

```diff
@@ -35,30 +32,28 @@ bool DrainManagerImpl::drainClose() const {
     return false;
   }
 
-  // We use the tick time as in increasing chance that we shutdown connections.
-  return static_cast<uint64_t>(drain_time_completed_.load()) >
-         (server_.random().random() % server_.options().drainTime().count());
-}
-
-void DrainManagerImpl::drainSequenceTick() {
-  ENVOY_LOG(trace, "drain tick #{}", drain_time_completed_.load());
-  ASSERT(drain_time_completed_.load() < server_.options().drainTime().count());
-  ++drain_time_completed_;
-
-  if (drain_time_completed_.load() < server_.options().drainTime().count()) {
-    drain_tick_timer_->enableTimer(std::chrono::milliseconds(1000));
-  } else if (drain_sequence_completion_) {
-    drain_sequence_completion_();
+  const MonotonicTime current_time = server_.dispatcher().timeSource().monotonicTime();
+  if (current_time >= drain_deadline_) {
+    return true;
   }
+
+  // P(return true) = elapsed time / drain timeout
+  const auto remaining_time =
+      std::chrono::duration_cast<std::chrono::seconds>(drain_deadline_ - current_time);
+  const auto elapsed_time = server_.options().drainTime() - remaining_time;
```

Comment:
is there a guarantee that drainTime() >= current_time ? Or should we do a compare first? I don't totally understand the surrounding draining architecture, but it looks like drainClose() can be called by something not controlled by the timer, so current_time_ may just be later. In that case maybe we should skip the subtraction underflow and compare to random and just return true?

Reply:
Assuming you mean drain_deadline_ (the computed deadline time point) and not drainTime() (the server configuration)? This case is tested for at [drain_manager_impl_test.cc:83](https://github.com/envoyproxy/envoy/pull/11138/files/3edaebebe35821a32b9a7b8b2112c3e0729633c9#diff-8878b807be959b98509feb61ed0452c7R82). If current_time > drain_deadline, the probability of drain close is 100%,

Reply:
sorry I meant checking server_.options().drainTime() >= remaining_time

or is that guaranteed to be true due to the wait remaining_time is calculated?

Reply:
(psuedo variables)

drain_deadline = drain_timeout + previous_time
remaining_time = drain_deadline - current_time
remaining_time = drain_timeout + previous_time - current_time
current_time >= previous time, therefore remaining_time <= drain_timeout

Math checks out to me!

Reply:
awesome. add assert then?

Reply:
Not sure if it's the exact ASSERT you wanted but 

drain >= remaining
drain - remaining >= 0
elapsed = drain - remaining
elapsed >= 0

so I added ASSERT(elapsed_time >= 0);

Reply:
probably not, if elapsed_time is ultimately a uint_something.

I meant asserting the subtraction is non-underflowing.
```
ASSERT(server_.options().drainTime() >= remaining_time);
```